### PR TITLE
Add a page describing the goals of Dune

### PR DESCRIPTION
**Pull request by jeremiedimino**
To give a clear direction to the project

### CODE REVIEW

File: doc/goals.rst [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+order to be such a tool, but we are slowly working towards that goal.
+On a practical level, there are a few boxes to check. There are in
+fact a ton details to sort out, but at a high-level a tool that works
+for everyone in the OCaml community should at least:
+
+1. have excelent backward compatibility properties
+2. have a robust and scalable core
+3. remain a no-brainer dependency
+4. have very good support for the OCaml language
+5. be extensible
```
Comment (shonfeder) [inline doc/goals.rst:23]: I just noticed that none of these items really speak to usability or accessibility. But this seems like an important goal for "a high-level a tool that works for everyone in the OCaml community".

File: doc/goals.rst [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+as well as people who work on their free time and cannot provide the
+same level of continuous support and updates.
+
+From the point of view of Dune, we have to assume that a released
+project with `dune` files is a precious piece that will potentially
+never change. So changing Dune in a way that it could no longer
+understand a released project is by default a no-no.
+
+Of cource, we can't give a 100% guarantee that Dune will always behave
+exactly the same. That would be irrealistic and would prevent the
```
Comment (shonfeder) [inline doc/goals.rst:47]:
```suggestion
exactly the same. That would be unrealistic and would prevent the
```

File: doc/goals.rst [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+features. But for now, Dune is a standalone build system that is
+primarily serving the needs of the OCaml community, and to the extent
+that is reasonably possible the needs of friend communities.
+
+Be extensible
+=============
+
+No matter how good the support for the OCaml language is, it will
+never be enough to cover every single project need. For this reason,
+Dune needs to provide some form of openness for projects that needs
```
Comment (shonfeder) [inline doc/goals.rst:130]:
```suggestion
Dune needs to provide some form of openness for projects that need
```

File: doc/goals.rst [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+something that don't completely fit in the model provided by Dune.
```
Comment (snowleopard) [inline doc/goals.rst:131]:
```suggestion
something that doesn't completely fit in the model provided by Dune.
```

Comment (shonfeder) [inline doc/goals.rst:131]:
```suggestion
something that doesn't completely fit in the model provided by Dune.
```

File: doc/goals.rst [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+In the long run, extensibility tends to get in the way of inovation
```
Comment (snowleopard) [inline doc/goals.rst:133]:
```suggestion
In the long run, extensibility tends to get in the way of innovation
```

File: doc/goals.rst [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+possible. Currently, the only requirement to build ``dune`` is a
+working OCaml compiler. Nothing else is required, not even a shell and
+we should keep it this way.
+
+Have very good support for the OCaml language
+=============================================
+
+There are many many build systems out there. What makes Dune different
+is that it targets the OCaml community primarily. So Dune must come
+with excelent support for the OCaml language and OCaml projects in
```
Comment (snowleopard) [inline doc/goals.rst:113]:
```suggestion
with excellent support for the OCaml language and OCaml projects in
```

File: doc/goals.rst [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+At this point, we have done a good job at 1, 3 and 4, we are working
+towards 2 and are doing the preparatory work for 5. Once all these
+boxes have been checked, we will consider that the Dune project will
+be feature complete.
+
+In the rest of this page, we develop these points and give some
+insights into our current and future focuses.
+
+Have excelent backward compatibility properties
```
Comment (snowleopard) [inline doc/goals.rst:33]:
```suggestion
Have excellent backward compatibility properties
```

File: doc/goals.rst [hunk 8]
```diff
@@ -0,0 +1,141 @@
+************
+Goal of dune
```
Comment (snowleopard) [inline doc/goals.rst:2]:
```suggestion
Goal of Dune
```

**Comment by jeremiedimino** [comment]
No problem, thanks @snowleopard and @shonfeder for your comments and fixes!

> Also, you sometimes write Dune and sometimes dune and I didn't spot a clear pattern to see why. Perhaps, you can just replace everything with Dune.

I usually use Dune to talk about the project and `dune` to talk about the binary. Maybe we should use explicitly "`dune` binary" or even "Dune binary" to refer to the binary, to avoid ambiguity.

File: doc/goals.rst (hunk 1 above)
Comment (jeremiedimino) [reply doc/goals.rst:23]: Indeed, you are right that this is an important point. I'll add something about it.

File: doc/goals.rst [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+communities such as Coq and possibly Reason/Bucklescript in the
+future.
+
+We haven't reached this goal yet and Dune still lacks in some areas in
+order to be such a tool, but we are slowly working towards that goal.
+On a practical level, there are a few boxes to check. There are in
+fact a ton details to sort out, but at a high-level a tool that works
+for everyone in the OCaml community should at least:
+
+1. have excelent backward compatibility properties
```
Comment (jberdine) [inline doc/goals.rst:19]:
```suggestion
1. have excellent backward compatibility properties
```

File: doc/goals.rst [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+project with `dune` files is a precious piece that will potentially
+never change. So changing Dune in a way that it could no longer
+understand a released project is by default a no-no.
+
+Of cource, we can't give a 100% guarantee that Dune will always behave
+exactly the same. That would be unrealistic and would prevent the
+project from moving forward.  In order for us to provide good backward
+compatibility properties while still keeping the project fresh and
+dynamic, we have to make sure to properly delimit, document and
+version the set of behaviors that users should relly on. And for this
```
Comment (jberdine) [inline doc/goals.rst:51]:
```suggestion
version the set of behaviors on which users should rely. And for this
```

File: doc/goals.rst [hunk 11]
```diff
@@ -0,0 +1,141 @@
+************
+Goal of Dune
+************
+
+The main goal of the Dune project is to provide the best possible
+build tool for the whole of the OCaml community, from individual
+developpers who work on open source projects during their free time
```
Comment (rgrinberg) [inline doc/goals.rst:7]: `developers`.

File: doc/goals.rst (hunk 10 above)
Comment (jeremiedimino) [reply doc/goals.rst:51]: Thanks for the typo fixes!

File: doc/goals.rst (hunk 11 above)
Comment (yminsky) [inline doc/goals.rst:7]: developers

Comment (jeremiedimino) [reply doc/goals.rst]: fixed

**Comment by jeremiedimino** [comment]
I added a section about accessibility. @shonfeder could you have a look? Anyone else is welcome to as well.

File: doc/goals.rst [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+possible. Currently, the only requirement to build Dune is a
+working OCaml compiler. Nothing else is required, not even a shell and
+we should keep it this way.
+
+Remain accessible
+=================
+
+Since Dune aims to be the best possible tool for the whole OCaml
+community, it is important for Dune to remain accessible. Getting
+started and leaning Dune should be straightforward.
```
Comment (avsm) [inline doc/goals.rst:114]:
```suggestion
started and learning Dune should be straightforward.
```

File: doc/goals.rst [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+Remain accessible
+=================
+
+Since Dune aims to be the best possible tool for the whole OCaml
+community, it is important for Dune to remain accessible. Getting
+started and learning Dune should be straightforward.
+
+For that purpose, when designing the language, the command line
+interface or the documentation, we must always put ourselves from the
+point of view of a user who is discovering Dune and its features.
```
Comment (shonfeder) [inline doc/goals.rst:118]: For a slightly more idiomatic phrasing, we might put something like

```suggestion
interface or the documentation, we must take on the perspective 
of a user who is just discovering Dune and its features.
```

Comment (jeremiedimino) [reply doc/goals.rst:118]: Thanks, you version is more succinct, I like it better :)

**Pull request merged**